libshaderc_spvc/src/spvcir_pass.cc

```diff
@@ -127,13 +194,241 @@ void SpvcIrPass::GenerateSpirvCrossIR(Instruction *inst) {
       break;
     }
 
+    // opcode: 16
+    case SpvOpExecutionMode: {
+      auto &execution = ir_->entry_points[inst->GetSingleWordInOperand(0u)];
+      auto mode =
+          static_cast<SpvExecutionMode>(inst->GetSingleWordInOperand(1u));
+      execution.flags.set(mode);
+
+      switch (mode) {
+        case SpvExecutionModeInvocations:
+          execution.invocations = inst->GetSingleWordInOperand(2u);
+          break;
+
+        case SpvExecutionModeLocalSize:
+          execution.workgroup_size.x = inst->GetSingleWordInOperand(2u);
+          execution.workgroup_size.y = inst->GetSingleWordInOperand(3u);
+          execution.workgroup_size.z = inst->GetSingleWordInOperand(4u);
+          break;
+
+        case SpvExecutionModeOutputVertices:
+          execution.output_vertices = inst->GetSingleWordInOperand(2u);
+          break;
+
+        default:
+          break;
+      }
+      break;
+    }
+
+    // opcode: 5
+    case SpvOpName: {
+      uint32_t id = inst->GetSingleWordInOperand(0u);
+      ir_->set_name(id, reinterpret_cast<const char *>(
+                            inst->GetInOperand(1u).words.data()));
+      break;
+    }
+      // opcode:6
+    case SpvOpMemberName: {
+      uint32_t id = inst->GetSingleWordInOperand(0u);
+      uint32_t member = inst->GetSingleWordInOperand(1u);
+      ir_->set_member_name(
+          id, member,
+          reinterpret_cast<const char *>(inst->GetInOperand(2u).words.data()));
+      break;
+    }
+
+    // opcode: 71
+    case SpvOpDecorate:
+    // opcode: 332
+    case SpvOpDecorateId: {
+      // spirv-cross comment:
+      // OpDecorateId technically supports an array of arguments, but our only
+      // supported decorations are single uint, so merge decorate and
+      // decorate-id here.
+      // TODO(sarahM0): investigate if this limitation is acceptable.
+      uint32_t id = inst->GetSingleWordInOperand(0u);
+
+      auto decoration =
+          static_cast<spv::Decoration>(inst->GetSingleWordInOperand(1u));
+      if (inst->NumInOperands() > 2) {
+        ir_->meta[id].decoration_word_offset[decoration] =
+            uint32_t(inst->GetInOperand(2u).words.data() - ir_->spirv.data());
```

Comment:
This doesn't look right.

Unfortunately SPIRV-Cross doesn't document what decoration_word_offset means.

In any case, you're taking the difference between pointers in two different allocated buffers.  The spirv-opt Instruction holds onto Operand objects, and each Operand has its own independently allocated vector of words, and the words.data() is pointing at the start of that buffer.

I think you want something like:

      offset_ /* the word offset of this instruction */  + 2